### PR TITLE
`int.__sizeof__` from #1426

### PR DESCRIPTION
**Pull request by ashermancinelli**
Impl from @RobertBerglund 's comment, working on #304

### CODE REVIEW

File: vm/src/obj/objint.rs [hunk 1]
```diff
@@ -186,6 +186,11 @@ impl PyInt {
         PyInt::new(options.get_int_value(vm)?).into_ref_with_type(vm, cls)
     }
 
+    #[pymethod(name = "__sizeof__")]
+    fn sizeof(&self, _vm: &VirtualMachine) -> usize {
+        (self.value.bits() as f64 / 8.).ceil() as usize
```
Comment (ashermancinelli) [inline vm/src/obj/objint.rs]: Results in:
```python
>>>>> for i in range(0, 100, 10):
..... 	print((i).__sizeof__())
..... 
0
1
1
1
1
1
1
1
1
1
```

File: vm/src/obj/objint.rs [hunk 2]
```diff
@@ -188,7 +189,7 @@ impl PyInt {
 
     #[pymethod(name = "__sizeof__")]
     fn sizeof(&self, _vm: &VirtualMachine) -> usize {
-        (self.value.bits() as f64 / 8.).ceil() as usize
+        mem::size_of::<Self>() as usize
```
Comment (ashermancinelli) [inline vm/src/obj/objint.rs]: Results in:
```python
>>>>> for i in range(0, 100, 10):
..... 	print((i).__sizeof__())
..... 
32
32
32
32
32
32
32
32
32
32
```

**Comment by ashermancinelli** [comment]
@coolreader18 
```rust
mem::size_of::<Self>() as usize
```
This'll just always return 32 I believe, I'm not sure how useful this is. You're right that it goes deeper than I thought :sweat_smile:

**Comment by coolreader18** [comment]
Yeah, it's not as simple as `mem::size_of::<T>()` because that just returns the size of `T` on the stack, which is not always necessarily a meaningful value. [This thread](https://users.rust-lang.org/t/solved-rust-round-usize-to-nearest-multiple-of-8/25549) from the rust forums might be useful; I'm not sure if there was some sort of bug with the f64 conversion.

**Comment by coolreader18** [comment]
Although I don't think there was, it just didn't go above 1 byte because 100 can be contained with just 8 binary digits; you'd have to go above 256 to see a change.

File: vm/src/obj/objint.rs [hunk 3]
```diff
@@ -189,7 +188,7 @@ impl PyInt {
 
     #[pymethod(name = "__sizeof__")]
     fn sizeof(&self, _vm: &VirtualMachine) -> usize {
-        mem::size_of::<Self>() as usize
+        self.value.bits() + 7 & !7
```
Comment (ashermancinelli) [inline vm/src/obj/objint.rs]:
```python
>>>>> (2**1).__sizeof__()
8
>>>>> (2**8).__sizeof__()
16
>>>>> (2**16).__sizeof__()
24
>>>>> (2**32).__sizeof__()
40
```

Comment (ashermancinelli) [reply vm/src/obj/objint.rs]: @coolreader18 I think you're right about just not going high enough, I'll test that out

Comment (youknowone) [reply vm/src/obj/objint.rs]: sizeof int is sum of size of int object itself + sum of its heap allocated memory.
So, roughly, `mem::sizeof<T> + 4 * self.value.data.len()` is expected.
If current code describes the latter, adding `mem::sizeof` will make the value correct

File: vm/src/obj/objint.rs [hunk 4]
```diff
@@ -188,7 +189,7 @@ impl PyInt {
 
     #[pymethod(name = "__sizeof__")]
     fn sizeof(&self, _vm: &VirtualMachine) -> usize {
-        self.value.bits() + 7 & !7
+        mem::size_of::<Self>() + (self.value.bits() + 7) & !7
```
Comment (ashermancinelli) [inline vm/src/obj/objint.rs]:
```python
40
>>>>> (2**8).__sizeof__()
48
>>>>> (2**16).__sizeof__()
56
>>>>> (2**32).__sizeof__()
72
```

Comment (coolreader18) [reply vm/src/obj/objint.rs]: `__sizeof__` should actually return bytes, so you'd probably want to divide the bits you get by 8

File: vm/src/obj/objint.rs [hunk 5]
```diff
@@ -576,7 +576,7 @@ impl PyInt {
 
     #[pymethod(name = "__sizeof__")]
     fn sizeof(&self, _vm: &VirtualMachine) -> usize {
-        size_of::<Self>()
+        (size_of::<Self>() + (self.value.bits() + 7) & !7) / 8
```
Comment (ashermancinelli) [inline vm/src/obj/objint.rs]: Divided by 8 to return bytes instead of bits.

Comment (coolreader18) [reply vm/src/obj/objint.rs]: `mem::size_of` actually does return bytes, it's the `BigInt::bits` method that returns bits.

File: vm/src/obj/objint.rs [hunk 6]
```diff
@@ -576,7 +576,7 @@ impl PyInt {
 
     #[pymethod(name = "__sizeof__")]
     fn sizeof(&self, _vm: &VirtualMachine) -> usize {
-        size_of::<Self>()
+        (size_of::<Self>() + ((self.value.bits() + 7) & !7)) / 8
```
Comment (coolreader18) [inline vm/src/obj/objint.rs]:
```suggestion
        size_of::<Self>() + ((self.value.bits() + 7) & !7) / 8)
```

**Pull request merged**